### ISSUE TITLE
False positive of -Wunused-local-typedef in dependent code.

### ISSUE THREAD

**Issue by hokein** [llvm/llvm-project#98878]
Given the following code:

```
#define TYPE_OR_NONTYPE  typename
template<TYPE_OR_NONTYPE T>
struct BaseT {
  static int ff();
  void Foo();
};

template<TYPE_OR_NONTYPE T>
struct DerivedT : BaseT<T> {
  void Inner();
};

template<TYPE_OR_NONTYPE T>
void DerivedT<T>::Inner() {
  using K = BaseT<T>; 
  this->K::Foo(); 
}
```

clang emits a unused typedef warning.

```
⇒  ./bin/clang -Xclang  -fsyntax-only -Wunused-local-typedef  --std=c++20 /tmp/t2.cpp
/tmp/t2.cpp:15:9: warning: unused type alias 'K' [-Wunused-local-typedef]
   15 |   using K = BaseT<T>;
      |  
``` 

This is shown in https://github.com/abseil/abseil-cpp/pull/1711.

**Comment by nico** [comment]
This doesn't repro for me, using a 3-week old clang:

`thakis@Nicos-MacBook-Pro llvm-project % cat test.cc`

```c++
#define TYPE_OR_NONTYPE  typename

template<TYPE_OR_NONTYPE T>
struct BaseT {
  static int ff();
  void Foo();
};

template<TYPE_OR_NONTYPE T>
struct DerivedT : BaseT<T> {
  void Inner();
};

template<TYPE_OR_NONTYPE T>
void DerivedT<T>::Inner() {
  using K = BaseT<T>; 
  this->K::Foo(); 
}
```

```
thakis@Nicos-MacBook-Pro llvm-project % out/gn/bin/clang  -Xclang  -fsyntax-only -Wunused-local-typedef  -std=c++20 test.cc
-macosx_version_min has been renamed to -macos_version_min
-macosx_version_min has been renamed to -macos_version_min
ld: library 'System' not found
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Am I holding it wrong? Does it require a very recent clang to repro?

**Comment by glandium** [comment]
I was definitely getting the warning on clang 18.

**Comment by glandium** [comment]
This can be reproduced with clang 18 with:

```c++
#include <utility>

#define TYPE_OR_NONTYPE typename
template<TYPE_OR_NONTYPE T>
struct BaseT {
  static int ff();
  void Foo();
};

template<TYPE_OR_NONTYPE T>
struct DerivedT : BaseT<T> {
  void Inner();
};

template<TYPE_OR_NONTYPE T>
void DerivedT<T>::Inner() {
  using K = BaseT<T>; 
  std::move(*this)->K::Foo(); 
}
```

**Comment by shafik** [comment]
This looks like what you mean I think: https://godbolt.org/z/7nrdeM535

```cpp
#define TYPE_OR_NONTYPE  typename
template<TYPE_OR_NONTYPE T>
struct BaseT {
  static int ff();
  void Foo();
};

template<TYPE_OR_NONTYPE T>
struct DerivedT : BaseT<T> {
  void Inner();
};

template<TYPE_OR_NONTYPE T>
void DerivedT<T>::Inner() {
  using K = BaseT<T>; 
  auto s = *this;
  s.K::Foo(); 
}
```

**Comment by b1tflyyyy** [comment]
Still reproducible on the current Clang version (22.1.0).
Note: I tried the same code with GCC, and I did not get this issue.

Used the example from [shafik](https://github.com/llvm/llvm-project/issues/98878#issuecomment-2325299506):
```cpp
#define TYPE_OR_NONTYPE  typename
template<TYPE_OR_NONTYPE T>
struct BaseT {
  static int ff();
  void Foo();
};

template<TYPE_OR_NONTYPE T>
struct DerivedT : BaseT<T> {
  void Inner();
};

template<TYPE_OR_NONTYPE T>
void DerivedT<T>::Inner() {
  using K = BaseT<T>; 
  auto s = *this;
  s.K::Foo(); 
}
```

Links:
[clang](https://godbolt.org/z/36KsP5hs3)
[clang-trunk](https://godbolt.org/z/j38YY4es9)
[gcc](https://godbolt.org/z/s7PGacs4z)
